### ISSUE TITLE
Play with a randomly-generated button

### ISSUE THREAD

**Issue by jimmosk** [buttonmen-dev/buttonmen#779]
The previous site had an option to create a game where the players had randomly-generated buttons, with different skill-densities ranging from no skills (though perhaps one Swing die?) up to what I think was termed the CRAZY level (most dice having one or more skills).  
Can someone recall what those levels of skill-density were? And more importantly, I'd officially like to request this functionality be added.
[edited title to make clear I didn't mean simply selecting an existing button randomly]

**Comment by irilyth** [comment]
Hey, so, yay! What's next on this ticket, come up with some other flavors of RandomBM? (Given "vanilla", should the others in fact be named after flavors too? Chocolate, Strawberry, CookieDough, Salmon, Sriracha... Ok, perhaps not. :^)

I'd be happy to help with figuring out a full set of flavors; or to just suggest a perhaps-obvious idea for the next flavor: Soldier (four regular dice plus one X swing).

**Comment by irilyth** [comment]
Huh, here is another random idea: We could make the RandomBM buttons based on other buttonsets. (Possibly even part of those buttonsets? Maybe that's not desirable, not sure.) So we could have RandomBM Vampyres, RandomBM Brom, RandomBM Foglio, etc, with the algorithm for creating them designed to give you results very much like the other buttons in those sets.

And then maybe other other things too, but that might be an interesting start. And I actually kind of like the idea of "give me a random button that looks like the other buttons in this set" for a wide range of sets.

Other thoughts?

**Comment by yawetag** [comment]
Didn't the old site have random buttons based on several sets? Or am I
mis-remembering?


Andrew Senger
asenger@gmail.com

**Comment by irilyth** [comment]
> Didn't the old site have random buttons based on several sets? Or am I mis-remembering?

It had different types, and they were sometimes labeled as being related to sets, but only a few. I happened to have written it down once, and I put a list in a comment early in this ticket. (Short summary: Soldiers, Fantasy/Foglio, and Anime. I think things like Brom, Brawl, L5R, Button Brains, Diceland, and Samurai, would also make good additions.)

**Comment by blackshadowshade** [comment]
For information, adding a new RandomBM type is close to trivial---and would be a great place for an up-and-coming back-end dev to jump into adding new button skills!

Pinging @nkmcalli, @ManvilleFog.  ;)

**Comment by nkmcalli** [comment]
Soon - almost there...

**Comment by jimmosk** [comment]
Okay, so I'm going to make my January BFotM adding another few types of RandomBM -- say, the Fixed and Mixed types listed above.  Or any others that you think would be easier or more fun.

**Comment by irilyth** [comment]
> Okay, so I'm going to make my January BFotM adding another few types of RandomBM -- say, the Fixed and Mixed types listed above.  Or any others that you think would be easier or more fun.

Ok! Here's a list of suggested types, based on button sets; all with five dice, for simplicity's sake:

- RandomBM BRAWL
  - 4 regular dice
  - one S or X swing
  - Speed on 1 - 2 dice

- RandomBM BROM
  - four regular dice
  - one V or X swing
  - Poison on 1 - 2 dice
  - Shadow on 0 - 2 dice

- RandomBM Button Brains
  - 3 - 4 regular dice
  - the other 1 - 2 dice with one of the following:
    - U, V, W, or X swing, with a 1/3 chance that it als has Mood
    - Twin
    - Option
  - Konstant on one die that is *not* a Swing, Twin, or Option die

- RandomBM Buttonlords
  - 4 regular dice
  - one X swing
  - one Auxiliary die, with a 1/4 chance that it is also an X swing
  - Shadow on 0 - 2 dice (including the Aux die)

- RandomBM Diceland
  - 3 - 4 regular dice
  - the other 1 - 2 X swing
  - Stinger on 1 - 2 dice
  - Poison on 0 - 1 *non*-Stinger dice (1/3 chance)

- RandomBM Fantasy
  - 1 - 3 regular dice
  - the other 2 - 4 Option dice

- RandomBM Legend of the Five Rings
  - 5 regular dice
  - Focus on 2 dice

- RandomBM Lunch Money
  - 2 - 4 regular dice with 1 - 8 sides
  - one X or Y swing
  - the other 0 - 2 dice with 10 - 20 sides, but at most one with 20 sides
  - Trip on 1 die with 1 - 6 sides, the smallest or second-smallest

- RandomBM Metamorphers
  - 4 regular dice
  - one X swing
  - Morphing on one die

- RandomBM Sailor Moon
  - 4 regular dice
  - 4 reserve dice

- RandomBM Soldiers
  - 4 regular dice
  - one X swing
  - no skills on any dice

- RandomBM Samurai
  - 3 - 4 regular dice
  - the other 1 - 2 V or X swing
  - Focus on one die

- RandomBM Studio Foglio
  - 3 regular dice
  - one Twin die, with a 1/2 chance that it has Poison
  - one W or X swing, with a 1/10 chance that it is *also* a Twin die

- RandomBM Vampyres
  - 4 regular dice
  - one X swing
  - Shadow on 1 - 2 dice (1/3 chance of 1, 2/3 chance of 2)

For the ranges (e.g. "3 - 4 regular dice"), I think an even chance of any of the outcomes is probably fine. If anyone wants to look more closely, and argue that a non-even chance would be more consistent with the buttons in the set, I'm totally fine with that, I just didn't bother most of the time.

Corrections and other feedback welcome too, of course.

If that sounds generally good: Jimmosk, pick your favorite, and maybe a few runners up? Or whoever takes the ticket can, of course.

**Comment by jimmosk** [comment]
Josh, you may not have noticed that 'Vanilla' was taken from your own copypasta in the first comment to this Issue -- it's one of the types of RandomBM from the previous site.  The two types which I requested as my BFotM are from your paste as well:
Fixed   5 Dice, no Swing, One skill on 2 Dice from cfd 
Mixed   5 Dice, no Swing, Three skills on 2 Dice each

**Comment by irilyth** [comment]
Sure, that's fine; I don't think those are all that interesting, because they're all a fixed set of dice with no swing or option, but as you like.

I also don't like "Fixed" and "Mixed" as names -- I don't think the names on the old site made much sense in general -- any suggestions?

**Comment by jimmosk** [comment]
Whereas I don't think Swing and Option are as interesting as the interplay between many different sorts of skills, which is why I want to make progress along that track. CRAZY, here we come!

**Comment by blackshadowshade** [comment]
Well, jimmosk, if you want CRAZY, then request CRAZY! There's no difference in difficulty between different types, since all we're doing is generating strings that conform to the spec, and then using all the standard machinery to parse them.

**Comment by irilyth** [comment]
> There's no difference in difficulty between different types, since all we're doing is generating strings that conform to the spec, and then using all the standard machinery to parse them.

Does the spec have a way to express things like "a 32% chance of five 20-siders, a 17% chance of four buttons with 3 - 14 sides plus two 20-siders, and a 51% chance of one of each type of swing die"? Or less nutty things, like "two skills, each on two dice, but no dice with more than one skill" vs "...and dice can have more than one skill"?

What *does* the spec say? :^)

**Comment by blackshadowshade** [comment]
We are defining the spec *here*, and once I have the spec, I will program to it.

**Comment by jimmosk** [comment]
Thanks, blackshadowshade, but I think introducing them "in order" to get people gradually acclimated to the possibility of CRAZYness is the way to go.  
To make the specs more, er, specific:

Fixed: Five dice, chosen from standard button sizes, with two of them having a single skill chosen (independently) from c, f, and d.

Mixed: Five dice, chosen from standard button sizes, with three skills chosen from all existing skills and each given to two of the five dice (independently, so some dice may get no skills and one or two could end up with all of them).

**Comment by blackshadowshade** [comment]
For those who are interested, I have a branch up with Fixed implemented. I won't submit a pull request until I have Mixed done, but since my Internet access is sporadic for the next few days, you may wish to see what is necessary to implement a new RandomBM type in the meanwhile.

**Comment by blackshadowshade** [comment]
I've just put in another two hours, and you can see the pull request that is up with both RandomBMFixed and RandomBMMixed implemented. Enjoy!

**Comment by jimmosk** [comment]
Should I add explanations of these RandomBM 'levels' to the Skills page? Or is that the wrong place since these aren't skills?

**Comment by irilyth** [comment]
I think it'd be good to have a separate page for RandomBM, since I think (hope!) there will be many types eventually.

I don't have a strong feeling about whether for now it should be on the dev wiki or the Wikia wiki. Eventually, it should presumably be on a page as part of the site itself...

**Comment by cgolubi1** [comment]
Hmm: http://dev.buttonweavers.com/ui/buttons.html?set=RandomBM has links to each of the buttons, and the individual button pages there describe each button.  I think that's the authoritative source.  How can we make it more useful?

**Comment by irilyth** [comment]
I figured we were talking about a single page that would describe all of the types. I don't know what Jimmosk had in mind.

**Comment by jimmosk** [comment]
I was talking about the same thing as Josh, and wondering if it should be on the same page as explains the skills. The page Chaos linked to does describe all the RandomBM types, but you have to look at each one separately. I think something like the condensed list Josh pasted in as the very first comment to this thread would be more useful than http://dev.buttonweavers.com/ui/buttons.html?set=RandomBM  is. 

Ideally, I think the consistent way is to put thumbnail descriptions of each type right into the pulldown menu, in parallel with (nearly?) every other button in them... is that doable?

**Comment by cgolubi1** [comment]
I agree that that would be useful, and it doesn't seem out of the question that if a button has a blank recipe, we could put something else in the pulldown.  Do you have in mind a format for thumbnail descriptions that would fit in the available space?

**Comment by irilyth** [comment]
Some feedback about the current implementation of RandomBMMixed on the dev site:

I think Slow should be removed from the skills list, because it's purely negative and not any fun. All the other skills, you might well say "ooh, my random button got Foo", and be excited about it, or "ooh, a die with both Foo and Bar, that should be interesting", or something positive; I don't see a situation where I would ever say "ooh, a die with Slow, yay".

Shadow and Maximum is a pretty funny combination. :^) Do we want to try to start a list of combinations to exclude? I don't know that sM is one of them, because it isn't totally useless, just a little weird; but there might be others where we'd say "oof, that is not a good idea" or some such. Maybe not, though; any thoughts? (In particular, do you want to hear about it on this ticket when weird combinations come up, or enh?)

**Comment by blackshadowshade** [comment]
In case anyone following this conversation hasn't noticed, on pull request #1500, we decided initially to exclude a number of skills from BMRandomMixed, because Chaos was hitting real bugs in her automated testing.

Currently, the excluded skills are:
- 'Auxiliary', 'Reserve', // game-level skills
- 'Berserk', 'Doppelganger', 'Mad', 'Mighty', 'Mood', 'Morphing', 'Radioactive', 'Weak', // recipe-changing skills
- 'Fire', // skills that add an extra step to attacks

We'd like to remove a number of skills from this list eventually, but we're also open to adding skills to this list, including Slow as proposed by irilyth.

**Comment by irilyth** [comment]
Jimmosk, you wrote up Fixed as

> Fixed: Five dice, chosen from standard button sizes, with two of them having a single skill chosen (independently) from c, f, and d.

My recollection is that on the old site, the skills were *not* chosen independently: Any given RandomBMFixed would have two dice with c, two with f, or two with d.

I don't feel strongly about it, and having them mixed here has turned up some interesting issues (like what happens with a button with both Chance and Focus), but if your goal was to faithfully recreate that types on the old site (for whatever reason), I don't think the current implementation does that.

**Comment by jimmosk** [comment]
Good catch!  Yes, I think it ought to be the same single skill, on two dice. My apologies for incorrectly parsing the original description. Blackshadowshade (or anyone), can you make that correction?

**Comment by blackshadowshade** [comment]
Sure, I can do that. Feel like fixing the description for me?

**Comment by jimmosk** [comment]
Fixed: Five dice, chosen from standard button sizes, with two of them having a single skill chosen from c, f, and d (the same skill on both).

**Comment by jimmosk** [comment]
Okay, it's a short month, so I'm going to use my February BFotM to request on what blackshadowshade said was "close to trivial---and would be a great place for an up-and-coming back-end dev to jump into adding new button skills!": I'd like the rest of the Old Site random buttons to be implemented, or as many as you can get to.  Here are their descriptions:

Soldier    4 Dice, chosen from standard button sizes, plus one X Swing, No Twin or Option Dice
Fantasy/Foglio    4 Dice, chosen from standard button sizes, plus one Swing of any type
Anime    4 Dice and 4 reserve dice, chosen from standard button sizes, Twins possible 
Standard    4 Dice, chosen from standard button sizes, plus one Swing of any type, with one skill chosen from all existing (non-excluded) skills and given to 2 Dice 
Multi     4 Dice, chosen from standard button sizes, plus one Swing of any type, with two skills chosen from all existing (non-excluded) skills and each given to 1 or more Dice 
Freaks    4 Dice, chosen from standard button sizes, plus one Swing of any type, with three skills chosen from all existing (non-excluded) skills and each given to 1 or more Dice 
CRAZY    4 Dice, chosen from standard button sizes, plus one Swing of any type, with five skills chosen from all existing (non-excluded) skills and each given to 1 or more Dice

**Comment by irilyth** [comment]
Jimmosk (or others), how would you feel about not using the old-site definitions for the ones that are like sets, but instead using the definitions I suggested earlier for those, which I think more closely match the other buttons in the sets?

**Comment by jimmosk** [comment]
Okay with me. You mean these ones?

Soldiers -- 4 regular dice - one X swing - no skills on any dice
Fantasy -- 1–3 regular dice - the other 2–4 Option dice
Studio Foglio -- 3 regular dice, one Twin die, with a 1/2 chance that it has Poison, one W or X swing, with a 1/10 chance that it is *also* a Twin die

**Comment by blackshadowshade** [comment]
Okay, let's take Anime first.

Die size distribution from Sailor Moon 1 and 2, excluding Wiseman and Shadow Warriors:

Normal die:
(1) : 1
(2) : 1
(4) : 12
(6) : 18
(8) : 8
(10) : 13
(12) : 19
(20) : 15

Reserve die:
(1) : 0
(2) : 2
(4) : 11
(6) : 12
(8) : 15
(10) : 19
(12) : 15
(20) : 16

Recipes are allowed to have repeated die sizes, so I'll generate recipes by sampling die sizes from a uniform distribution for 4 normal and 4 reserve dice.

**Comment by irilyth** [comment]
Jimmosk: Yep, that's what I had in mind. Also Anime, which I think is already done. :^) (The only difference there was about Twin dice; I think Cammy Neko and Sailor Main are the only ones with Twin dice, and they're not part of the main anime sets, so I think it'd be better to leave Twin dice out, or maybe give them a very small chance of showing up if that's easy to do and people like it better.)

**Comment by irilyth** [comment]
Should Tenchi Muyo count for distribution purposes? Probably doesn't matter a whole lot.

**Comment by jimmosk** [comment]
I like the possibility of them having twin dice, but only slightly, and primarily for reasons of more-possibilities-means-more-fun. And if blackshadowshade has already crafted the logic without having twins, I don't see any need to add them (just remove "Twins possible" from their description).

As for Tenchi Muyo, I don't see why not. They're anime buttons.

**Comment by jimmosk** [comment]
Any thoughts on whether the text after a game ends should be changed for RandomBM, specifically the "play with the same buttons" options? It implies the exact same random buttons, but that's not what you get...

**Comment by edpegg** [comment]
At http://www.buttonweavers.com/ui/forum.html#!threadId=96  , I started a list of possible skill pairs.  Some skill pairs don't make any sense, and I marked them with an X. 

In http://www.buttonweavers.com/ui/game.html?game=4594  , the RandomBMMixed button got a "cw(1)" die, a Slow Chance die.  

RandomBMMixed should work with a list of forbidden skill pairs.

**Comment by irilyth** [comment]
I think we've concluded that cw and c(1) are useless, but not forbidden. (There are definitely buttons that have Chance and Focus on dice that also have skills which prevent those dice from counting for initiative.)

**Comment by irilyth** [comment]
This is my BFOTM for March.

I don't care as much about the types from the old site. :^) Here's my preference list, of the set-based types I suggested earlier, with point values:

- 3: RandomBM Soldiers
- 3: RandomBM BROM
- 3: RandomBM Studio Foglio
- 2: RandomBM Samurai
- 2: RandomBM BRAWL
- 1: RandomBM Diceland
- 1: RandomBM Lunch Money
- 1: RandomBM Fantasy
- 1: RandomBM Legend of the Five Rings
- 0: RandomBM Vampyres
- 0: RandomBM Buttonlords
- 0: RandomBM Metamorphers
- 0: RandomBM Button Brains

I'd like as many points worth as can be done, so if one of the three-point ones is particularly hard, feel free to skip it if that means getting more points worth of other things done. Or ignore the whole points thing and just go in the order I listed. :^)

Two other notes:

* I think we already got "RandomBM Sailor Moon" from my list, in the form of RandomBM Anime.

* To recap the remaining types from the old site (which I don't care as much about, but so there's a concise recap here):
  * "Soldier" is covered by what I called "RandomBM Soldiers"
  * "Fantasy/Foglio" is covered by some combination of what I called "RandomBM Studio Foglio" and "RandomBM Fantasy" (which I think are pretty different, so I don't know why they were lumped together.
  * That leaves Standard, Multi, Freaks, and CRAZY still to do at some point.

Jimmosk, if you have an opinion about this, maybe chime in about what you'd like "each given to 1 or more Dice" to mean for the latter three? In particular, should each skill be given to 1 - 5 of the dice, with the number selected randomly and independently? Or should each skill end up on the same number of dice? Or should there be a consistent number of skill-on-dice total, but randomly as to how they're distributed? (e.g. for Multi, say a total of five skills-on-dice, so you might end up with one skill on one die and one on four, or one on two and one on three.) I sort of like that last one, and suggest maybe 3N-1 as a formula, i.e.:

- Standard: One skill, two skills-on-dice, so things like a(4) (4) a(6) (10) (X).
- Multi: Two skills, five skills-on-dice, so things like a(4) b(4) ab(6) a(10) (X). 
- Freaks: Three skills, eight skills-on-dice, so things like  ac(4) b(4) ab(6) ab(10) c(X). 
- CRAZY: Five skills, fourteen skills-on-dice, so things like  acd(4) be(4) abe(6) abde(10) cd(X). 

That also seems to beg for the addition of something like

- Wacky: Four skills, eleven skills-on-dice, so things like  acd(4) b(4) ab(6) abd(10) cd(X). 

Whatcha think? (If you like the general idea: 3N-1, or 2N? 2N might be a little less out there. Or it doesn't have to be a consistent formula. :^)

**Comment by jimmosk** [comment]
When I originally wrote my understandings of the terms (just after I put the bugfix of the month label on this 21 days before the day I'm writing this), I was thinking along the lines of your first version, with random, independently-chosen numbers of skills. But thinking about it some more, that could lead to really lopsided battles between random dice of the same type -- more than the lopsided ones that already happen, I mean.  So it'd probably be more fun to follow your 3N-1 plan. And I can't actually remember which is truer to the previous ButtonMen site. Can anyone else?

**Comment by irilyth** [comment]
Yeah, I have no idea what the previous site did. I sort of recall at least some of the descriptions not matching what actually happened. :^P  (Like, it said "five skills" for CRAZY, but you only ever got three skills, but all five *dice* had at least one skill, or something like that.) But that might have been due to a mis-reading of the descriptions, who knows.

**Comment by irilyth** [comment]
You know, on further reflection, I kind of like the idea of the Standard/Multi/Freaks/Wacky/Crazy ones as the next target.

One question: Should we change the names, to avoid any hint of pejorative terms for mental illness? I'm happy to try to come up with something else.

**Comment by jimmosk** [comment]
I'm in support of making that change. I think Crazy is the only one seen as pejorative, right?

**Comment by irilyth** [comment]
I think "Freaks" isn't great either; and I'm not all that enthused about "standard" and "multi", just because they don't seem all that meaningful.

How about some combination of: Hyper, super, ultra, uber, and/or mega? Perhaps combined with "skill", so RandomBM SingleSkill, RandomBM MultiSkill, RandomBM HyperSkill, RandomBM SuperSkill, RandomBM UltraSkill?

Or we could use actual numerical prefixes: Monoskill, Duoskill, Triskill, Tetraskill, Pentaskill. (Or more Latinesque: Uniskill, Biskill, Triskill, Quadroskill, Quintaskill. I like the Greek ones better.)

Whatcha think?

**Comment by jimmosk** [comment]
I'm not big on the first option, since Hyper etc don't have as much of an ordering in my head as Freaks etc (though I admit that may just be repeated exposure to the latter). 
I'm fine with the other two suggestions, and I too prefer the Greek.

**Comment by irilyth** [comment]
Ok, let's go with Greek! Any opinions about the root? I think "skill" is ok, something else could also work, I don't have any great ideas. I like not camelcasing it, so "RandomBM Tetraskill" etc.

**Comment by jimmosk** [comment]
I think skill is totally the proper root, since that's what we're randomly choosing one (or more) of. But it occurs to me that the Create Game menu lists buttons in alphabetical order within their set, which would be suboptimal here.  So either we jigger that around, or go with Random 1-skill, Random 2-skill, etc, methinks.

**Comment by irilyth** [comment]
Hmm. Dev folks, can the sorting of the buttons within a set, in the Create Game menu, be overridden? Or must it always be sorted by name?

**Comment by irilyth** [comment]
> We'd like to remove a number of skills from this list eventually, but we're also open to adding skills to this list, including Slow as proposed by irilyth.

Any objections to removing Slow?

**Comment by blackshadowshade** [comment]
irilyth, we explicitly specify the order of buttons in data.button.sql, so we have have whatever order we like.

**Comment by irilyth** [comment]
> irilyth, we explicitly specify the order of buttons in data.button.sql, so we have have whatever order we like.

Ok, great! Here's my request for the next set of RandomBM buttons, then:

* RandomBM Monoskill: Four regular dice, one R S T U V W X Y or Z swing; 1 skill, appearing a total of 2 times on various dice.
* RandomBM Duoskill: As above, but 2 skills, appearing a total of 5 times on various dice.
* RandomBM Triskill: As above, but 3 skills, appearing a total of 7 times on various dice.
* RandomBM Tetraskill: As above, but 4 skills, appearing a total of 10 times on various dice.
* RandomBM Pentaskill: As above, but 5 skills, appearing a total of 13 times on various dice.

I think 3N-1 is too many, but 2N might be too few, so I made up something in the middle here. (approximately 2.5 x N, I guess. :^)

I think we should also:

* Add Slow to the excluded list.
* Update the description for RandomBM Mixed to list the actual skills selected from (or the excluded skills), since right now it says "all existing skills".
* Have those five new ones show up at the end of the set, in that order.
* Have the current set sorted differently: Vanilla, Anime, Fixed, Mixed, would be my vote. (And then these new ones after that.)

Thanks!

**Comment by edpegg** [comment]
Over at Create Game, one option is Die Skill.  It lists all the various skills, ANY, Auxiliary, Berserk, ..., Warrior, and Weak.  

If one of these skills is selected, say Value, then Random Button, Beatnik Turtle, Miser, Noeh, schwa, The Fool are shown. The Random Button pretty much ignores the previous selection.

I propose that the prior selections mean something.  If I clicked Value, it would instead show
Random Existing Button + v (Value) (5 choices)
Beatnik Turtle, Miser, Noeh, schwa, The Fool 
Random Created Button + v (Value)

Second, add a few more dropdowns.  
One would be Die Skill, again.  This would allow a button to be specified with two skills.  
Value + Trip would give:
Random Existing Button + t v (Trip, Value) (2 choices)
Noeh, schwa
Random Created Button + t v (Trip, Value)
The second would be Number of Skills.  0, 1, 2, 3, 4, 5, 6, 7+.  This would be the number of distinct skills. For example, Da Pimp has 5 distinct skills.  

Currently, there are 3 dropdowns: Button set, Tournament legal, Die skill.  
I'm proposing 5 dropdowns: Button set, Tournament legal, Die skill, Die skill, Number of skills.

With these, I could create a random button with Shadow, Konstant, and 1 other skill (3 skills total).
,

**Comment by irilyth** [comment]
A randomly *generated* button isn't the same as a randomly selected button. Check out #1234 instead.

**Comment by irilyth** [comment]
Ah, I didn't read that whole comment carefully.

I think being able to make up on-the-fly formulas for a randomly-generated button would be a fine thing, but I think it's beyond the scope of this ticket. But it'd be a fine thing to make a new ticket for!

**Comment by jimmosk** [comment]
In http://dev.buttonweavers.com/ui/game.html?game=1745 I got a die with Speed and Stealth. But since Stealth dice can make no kind of attack other than multi-die Skill attacks, I can't actually use its Speed skill. That strikes me as dumb, and I'm wondering if Random buttons shouldn't be able to have Stealth paired with skills that are supposed to enable it to make other kinds of attacks (Shadow, Trip, etc).  I'm saying so here, but if you think it should be its own separate issue just let me know.

(Oh. and has anybody brought up the potential confusion of using 'skill' to mean a specific kind of attack as well as any of a large number of powers?)

**Comment by irilyth** [comment]
I had BFOTM-ed this previously (last March), and said

Here's my request for the next set of RandomBM buttons, then:

* RandomBM Monoskill: Four regular dice, one R S T U V W X Y or Z swing; 1 skill, appearing a total of 2 times on various dice.
* RandomBM Duoskill: As above, but 2 skills, appearing a total of 5 times on various dice.
* RandomBM Triskill: As above, but 3 skills, appearing a total of 7 times on various dice.
* RandomBM Tetraskill: As above, but 4 skills, appearing a total of 10 times on various dice.
* RandomBM Pentaskill: As above, but 5 skills, appearing a total of 13 times on various dice.

I think 3N-1 is too many, but 2N might be too few, so I made up something in the middle here. (approximately 2.5 x N, I guess. :^)

I think we should also:

* Add Slow to the excluded list.
* Update the description for RandomBM Mixed to list the actual skills selected from (or the excluded skills), since right now it says "all existing skills".
* Have those five new ones show up at the end of the set, in that order.
* Have the current set sorted differently: Vanilla, Anime, Fixed, Mixed, would be my vote. (And then these new ones after that.)

I don't think any of those actually got implemented, so that's still what I'd like to see here, and I'm BFOTM-ing it again for Jan 2016.

**Comment by blackshadowshade** [comment]
The description has already been updated, as far as I can see.

**Comment by blackshadowshade** [comment]
Slow has already been excluded.

**Comment by blackshadowshade** [comment]
The set ordering was changed, but apparently the button display algorithm always sorts them in alphabetical order within sets.

**Comment by blackshadowshade** [comment]
BTW, these changes happened in #1579, which partially addressed (i.e., two hours) the BotM of March last year.

**Comment by irilyth** [comment]
Ah, ok. As it turns out, I only barely cared about any of those things, but it's good that they're done. :^)

Is it feasible to get some of the new button types I mentioned in this current BFOTM?

**Comment by blackshadowshade** [comment]
Sure, shouldn't be too difficult.

**Comment by blackshadowshade** [comment]
Okay, I've found the code that orders the buttons:

https://github.com/buttonmen-dev/buttonmen/blob/master/src/engine/BMInterface.php

BMInterface->execute_button_data_query() has the following SQL ordering:

    ORDER BY s.sort_order ASC, b.name ASC;

Thus, it sorts by set order and then by name.

If you want the RandomBM buttons sorted in a particular order, we might have to do that by choosing names that sort in the correct order when sorted alphabetically. What do you think about calling the new ones something like RandomBM_with_1_skill, RandomBM_with_2_skills? It's not pretty, but it'll sort correctly.

**Comment by blackshadowshade** [comment]
The other option is that we order within sets by ID number, which maintains the ordering in the database. Any thoughts about that?

**Comment by AdmiralJota** [comment]
Or add a button.sort_order column, then `ORDER BY s.sort_order ASC, b.sort_order ASC, b.name ASC;`?

**Comment by blackshadowshade** [comment]
Good point. However, I'm still interested in whether people would like to have buttons displayed in canonical order within each set, as opposed to alphabetical order. For example, when I look at Brom, I expect to see Echo last, instead of in the middle of the set.

**Comment by dwvanstone** [comment]
Where does the canonical order come from? The way they were listed on the DHS site?

**Comment by irilyth** [comment]
I think that being able to control the ordering is good, so we don't just get stuck with alphabetical (or whatever). I don't have a strong preference about how we control it, or what it should be for sets other than RandomBM.

**Comment by AdmiralJota** [comment]
I don't have strong feelings about whether buttons are normally in alphabetical or canonical order, but I like being able to support it on a case-by-case basis with the sort_order field (so Fanatics can be alphabetical without anyone needing to manually set their sort orders individually, but RandomBM's can be canonical).

**Comment by blackshadowshade** [comment]
I was taking the canonical order as "the way the designer decided to list the buttons". In the case of the standard buttons, many of them are ordered in alphabetical order, but Soldiers isn't. Also, promos that were later included in a set (e.g., Echo) often don't show up in the canonical ordering, and were often listed on the DHS site after the standard set (if I remember correctly).

I'm happy to add a sort order for special sets where the sort order is important. For example, there's a Zodiac set currently being designed, and I think it would make sense to order those by month order, instead of alphabetical order.
